discretize/utils/meshutils.py

```diff
@@ -255,3 +257,438 @@ def ExtractCoreMesh(xyzlim, mesh, meshType='tensor'):
+        for dim in range(base_mesh.dim):
+
+            cc_base = getattr(
+                    base_mesh,
+                    "vectorCC{orientation}".format(
+                            orientation=axis[dim]
+                        )
+            )
+
+            cc_local = getattr(
+                    mesh,
+                    "vectorCC{orientation}".format(
+                            orientation=axis[dim]
+                        )
+            )
+
+            shift = (
+                cc_base[np.max([np.searchsorted(cc_base, center[dim])-1, 0])] -
+                cc_local[np.max([np.searchsorted(cc_local, center[dim])-1, 0])]
+            )
+
+            x0[dim] += shift
+
+            mesh.x0 = np.hstack(x0)
+
+    return mesh
+
+
+def refine_tree_xyz(
+            mesh, xyz,
+            method="radial",
+            octree_levels=[1, 1, 1],
+            octree_levels_padding=None,
+            finalize=False,
+):
+    """
+    Refine a TreeMesh based on xyz point locations
+
+    :param BaseMesh mesh: The TreeMesh object to be refined
+    :param numpy.ndarray xyz: 2D array of points
+    :param str method: Method used to refine the mesh based on xyz locations
+
+        - "radial": Based on radial distance xyz and cell centers
+        - "surface": Along triangulated surface repeated vertically
+        - "box": Inside limits defined by outer xyz locations
+    :param list octree_levels: Minimum number of cells around points
+        in each k octree level [N(k), N(k-1), ...]
+    :param list octree_levels_padding: Padding cells
+        added to the outer limits of the data each octree levels
+        used for method= "surface" and "box" [N(k), N(k-1), ...].
+    :param bool finalize: True | [False]    Finalize the TreeMesh
+    :return: TreeMesh mesh
+    """
+
+    if octree_levels_padding is not None:
+
+        if len(octree_levels_padding) != len(octree_levels):
+            raise Exception(
+                "'octree_levels_padding' must be the length %i" % len(octree_levels)
+            )
+
+    else:
+        octree_levels_padding = np.zeros_like(octree_levels)
+
+    # Prime the refinement against large cells
+    mesh.insert_cells(
+        xyz,
+        [mesh.max_level - np.nonzero(octree_levels)[0][0]]*xyz.shape[0],
+        finalize=False
+    )
+
+    # Trigger different refine methods
+    if method == "radial":
+
+        # Build a cKDTree for fast nearest lookup
+        tree = cKDTree(xyz)
+
+        # Compute the outer limits of each octree level
+        rMax = np.cumsum(
+            mesh.hx.min() *
+            np.asarray(octree_levels) *
+            2**np.arange(len(octree_levels))
+        )
+
+        # Radial function
+        def inBall(cell):
+            xyz = cell.center
+            r, ind = tree.query(xyz)
+
+            for ii, nC in enumerate(octree_levels):
+
+                if r < rMax[ii]:
+
+                    return mesh.max_level-ii
+
+            return 0
```

Comment:
could it be good to add a `min_level` parameter? so that you'll always get cells at least at that level? Maybe something that gets called first before any of these three get evaluated? A simple `mesh.refine(min_level)` should work.